src/host/rcp_host.cpp

```diff
@@ -364,10 +365,18 @@ void RcpHost::HandleStateChanged(otChangedFlags aFlags)
 
     mThreadHelper->StateChangedCallback(aFlags);
 
-    if ((aFlags & OT_CHANGED_THREAD_ROLE) && IsAttached() && mJoinReceiver != nullptr)
+    if (aFlags & OT_CHANGED_THREAD_ROLE)
     {
-        otbrLogInfo("Join succeeded");
-        SafeInvokeAndClear(mJoinReceiver, OT_ERROR_NONE, "Join succeeded");
+        otDeviceRole role = GetDeviceRole();
+        for (auto &callback : mThreadRoleChangedCallbacks)
+        {
+            callback(role);
+        }
+        if (IsAttached() && mJoinReceiver != nullptr)
+        {
+            otbrLogInfo("Join succeeded");
+            SafeInvokeAndClear(mJoinReceiver, OT_ERROR_NONE, "Join succeeded");
+        }
```

Comment:
```suggestion

        for (auto &callback : mThreadRoleChangedCallbacks)
        {
            callback(role);
        }

        if (IsAttached() && mJoinReceiver != nullptr)
        {
            otbrLogInfo("Join succeeded");
            SafeInvokeAndClear(mJoinReceiver, OT_ERROR_NONE, "Join succeeded");
        }
```
Nit: add a blank between blocks